### ISSUE TITLE
[Clang][TSan] Spurious read of std::nullptr_t member storage causes false-positive data race

### ISSUE THREAD

**Issue by hubert-reinterpretcast** [llvm/llvm-project#214922]
Clang miscompiles trivial copy/move construction/assignment for non-union class types containing `std::nullptr_t` members when TSan is enabled. In particular, it emits a call to `llvm.memcpy.p0.p0.i64` with arguments that read even the bytes of memory locations occupied by `std::nullptr_t` objects. The member-wise copy specified in [[class.copy.assign] paragraph 12](https://wg21.link/class.copy.assign#12) does not imply anything for accessing the source `std::nullptr_t` member different from lvalue-to-rvalue conversion, which does not access the memory location ([[conv.lval] paragraph 3](https://wg21.link/conv.lval#3)).

**Compiler Explorer:** https://godbolt.org/z/Y8bx3KE63

### Source (`/app/example.cpp`)
```cpp
#include <thread>
struct A {
  decltype(nullptr) p;
} b;
int main(void) {
  A a;
  std::thread t([&] {
    for (int i = 0; i < 128; ++i) {
      a.p = 0;
    }
  });
  for (int i = 0; i < 128; ++i) {
    b = a;
  }
  t.join();
}
```

### Compiler invocation
```bash
clang++ -std=c++26 -fsanitize=thread /app/example.cpp
```

### Compiler output (actual and expected)
(clean compile)

### Run invocation and output
```console
$ ./a.out
==================
WARNING: ThreadSanitizer: data race (pid=1)
  Read of size 8 at 0x7fffffffe990 by main thread:
    #0 __tsan_memcpy /root/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_memintrinsics.cpp:27:3 (output.s+0x607de)
    #1 main /app/example.cpp:13:7 (output.s+0xe9640)

  Previous write of size 8 at 0x7fffffffe990 by thread T1:
    #0 main::$_0::operator()() const /app/example.cpp:9:11 (output.s+0xe9c3d)
    #1 void std::__invoke_impl<void, main::$_0>(std::__invoke_other, main::$_0&&) /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0/../../../../include/c++/17.0.0/bits/invoke.h:63:14 (output.s+0xe9bd5)
    #2 std::__invoke_result<main::$_0>::type std::__invoke<main::$_0>(main::$_0&&) /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0/../../../../include/c++/17.0.0/bits/invoke.h:98:14 (output.s+0xe9b45)
    #3 void std::thread::_Invoker<std::tuple<main::$_0>>::_M_invoke<0ul>(std::_Index_tuple<0ul>) /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0/../../../../include/c++/17.0.0/bits/std_thread.h:303:13 (output.s+0xe9afd)
    #4 std::thread::_Invoker<std::tuple<main::$_0>>::operator()() /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0/../../../../include/c++/17.0.0/bits/std_thread.h:310:11 (output.s+0xe9aa5)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<main::$_0>>>::_M_run() /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0/../../../../include/c++/17.0.0/bits/std_thread.h:255:13 (output.s+0xe9969)
    #6 <null> <null> (libstdc++.so.6+0xf9585) (BuildId: 107b21dc20d33c8591215fed2771491c02556e91)

  Location is stack of main thread.

  Thread T1 (tid=3, finished) created by main thread at:
    #0 pthread_create /root/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:1082:3 (output.s+0x653da)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State>>, void (*)()) <null> (libstdc++.so.6+0xf9658) (BuildId: 107b21dc20d33c8591215fed2771491c02556e91)
    #2 main /app/example.cpp:7:15 (output.s+0xe9619)

SUMMARY: ThreadSanitizer: data race /app/example.cpp:13:7 in main
==================
ThreadSanitizer: reported 1 warnings
```

### Compiler version info (`clang++ -v`)
```
clang version 24.0.0git (https://github.com/llvm/llvm-project.git ceaff22ed2c3ceb84da3e9334f511bc3f5b757cd)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /cefs/62/62ce70ab7c0e0c4ed12dbff9_clang-assertions-trunk-20260807/bin
Build config: +assertions
Found candidate GCC installation: /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0
Selected GCC installation: /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/17.0.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
```

**Comment by hubert-reinterpretcast** [comment]
@zygoloid, thoughts? Should Clang (when TSan is enabled) adjust its codegen strategy for trivial copy/move to avoid reads of `std::nullptr_t` member storage?

fyi: @AaronBallman @cor3ntin @shafik

**Comment by zygoloid** [comment]
Why are we generating any code for the `a.p = 0;` assignment? I think that should probably not touch memory.

The memcpy intrinsic is given a description of which bytes are copied and which bytes are padding. Are we describing the nullptr_t as being padding? That seems like the proper fix here.

**Comment by hubert-reinterpretcast** [comment]
> Why are we generating any code for the `a.p = 0;` assignment? I think that should probably not touch memory.

TMK, the standard does not have wording to acknowledge that no access occurs when storing to a `std::nullptr_t` object. At least for writing via a volatile glvalue, we may need to touch the memory:
```cpp
int main(void) {
  bool t = true;
  volatile decltype(nullptr) a;
  while (t)
    a = nullptr; // prevents UB
}
```

https://wg21.link/expr.assign#2 plainly states that the object referred to by the left operand _is_ modified, albeit with elaboration upon _how_ the object is modified.

**Comment by hubert-reinterpretcast** [comment]
> The memcpy intrinsic is given a description of which bytes are copied and which bytes are padding. Are we describing the nullptr_t as being padding? That seems like the proper fix here.

We aren't describing the `nullptr_t` as padding and the `!tbaa.struct` metadata is not generated at `-O0`. So it seems a fix would require
- generating the metadata at `-O0` when TSan is enabled and
- describing `nullptr_t` members as padding.

**Comment by hubert-reinterpretcast** [comment]
Updating the front end to emit the `!tbaa.struct` metadata with the `nullptr_t` as padding will not be sufficient.

Given

```llvm
; Function Attrs: mustprogress nounwind sanitize_thread
define dso_local void @_ZN1A6assignERKS_(ptr noundef nonnull align 8 dereferenceable(8) %this, ptr noundef nonnull align 8 dereferenceable(8) %o) local_unnamed_addr #0 {
entry:
  call void @llvm.memcpy.p0.p0.i64(ptr align 8 %this, ptr align 8 %o, i64 8, i1 false), !tbaa.struct !0
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.memcpy.p0.p0.i64(ptr noalias writeonly captures(none), ptr noalias readonly captures(none), i64, i1 immarg) #1

attributes #0 = { mustprogress nounwind sanitize_thread }
attributes #1 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }

!0 = !{}
```

At `-O0`, `ThreadSanitizerPass` generates

```llvm
  %1 = call ptr @__tsan_memcpy(ptr %this, ptr %o, i64 8)
```

despite the metadata indicating padding.

At `-O1`, `InstCombinePass` would replace

```llvm
  call void @llvm.memcpy.p0.p0.i64(ptr align 8 %this, ptr align 8 %o, i64 8, i1 false), !tbaa.struct !0
```

with

```llvm
  %0 = load i64, ptr %o, align 8
  store i64 %0, ptr %this, align 8
```

It looks like a solution based on the `!tbaa.struct` metadata would require some changes in `opt`. @vitalybuka @MaskRay, thoughts?

**Comment by zygoloid** [comment]
> TMK, the standard does not have wording to acknowledge that no access occurs when storing to a `std::nullptr_t` object.

Ah, yeah. Looks like we discussed this at length on the core reflector starting June 13, 2019 (subject: "value representation of nullptr_t"). Seemed to be good consensus amongst CWG regulars that `nullptr_t` should have only padding bits, but it doesn't look like a core issue got created. May be worth asking Jens to open one.

> It looks like a solution based on the `!tbaa.struct` metadata would require some changes in `opt`.

We get the same tsan error for an equivalent case without `nullptr_t`: https://godbolt.org/z/9z53EGs51 (I couldn't get the padding bits to be zeroed with `new (this) Nullptr()`, not sure why -- the AST indicates that Clang knows the `CXXConstructExpr` is `zeroing`. Maybe we don't zero tail padding? Looks like a `CodeGen` bug to me.)

Do we expect races between stores to padding and copies of padding to be reported by TSan in general? My expectation is that a defaulted special member should behave like it copies the individual members, not like a memcpy, so I think this too is a false positive. (Even though some badly-behaved program might be storing state in the padding bits.)

**Comment by hubert-reinterpretcast** [comment]
> We get the same tsan error for an equivalent case without `nullptr_t`: https://godbolt.org/z/9z53EGs51 (I couldn't get the padding bits to be zeroed with `new (this) Nullptr()`, not sure why -- the AST indicates that Clang knows the `CXXConstructExpr` is `zeroing`. Maybe we don't zero tail padding? Looks like a `CodeGen` bug to me.)

Note that `-O` (or higher) is currently necessary for the `!tbaa.struct` metadata (not that it resolves the issue, but it is needed if the intent is to interrogate the IR transformations): https://godbolt.org/z/vPoroh5h4 (which uses the assignment operator from the thread as well, avoiding the explicit call to `memset`)

As for zeroing tail padding, I agree that we should be doing so in contexts where we are not expecting a potentially-overlapping subobject.

**Comment by hubert-reinterpretcast** [comment]
> Do we expect races between stores to padding and copies of padding to be reported by TSan in general? My expectation is that a defaulted special member should behave like it copies the individual members, not like a memcpy, so I think this too is a false positive. (Even though some badly-behaved program might be storing state in the padding bits.)

Such a "badly-behaved" program should be using `memcpy` explicitly, which AFAICT, does not attach the `!tbaa.struct` metadata. I would expect that "copies of padding" via such a call can contribute to a data race under the standard (at least where there is an in-lifetime array providing storage for the object that has the padding). I think we should limit the question to cases where the `!tbaa.struct` metadata is attached. When such metadata is attached, the claim seems to be that TSan should not consider the padding to be accessed. An additional question is whether (apart from TSan) the accesses introduced by the implementation violate the semantics of the abstract machine.
